File: code/appendix/torch_tansacnet/OrthonormalMatrixGenerationSystem.py
import torch
import math

def rot_(vt, vb, angle):
    """ Planar rotation """
    c = torch.cos(angle)
    s = torch.sin(angle)
    u = s * (vt + vb)
    vt = (c + s) * vt
    vb = (c - s) * vb
    vt = vt - u
    vb = vb + u
    return vt, vb

class OrthonormalMatrixGenerationSystem:
    """
    ORTHONORMALMATRIXGENERATIONSYSTEM Orthonormal matrix generator
    
    Requirements: Python 3.10/11.x, PyTorch 2.3.x
    
    Copyright (c) 2024, Shogo MURAMATSU
    
    All rights reserved.
    
    Contact address: Shogo MURAMATSU,
                    Faculty of Engineering, Niigata University,
                    8050 2-no-cho Ikarashi, Nishi-ku,
                    Niigata, 950-2181, JAPAN
    
    http://www.eng.niigata-u.ac.jp/~msiplab/
    """

    def __init__(self,
                 dtype=torch.get_default_dtype(),
                 partial_difference=False,
                 mode='normal'):
        super(OrthonormalMatrixGenerationSystem, self).__init__()
        self.dtype = dtype
        self.partial_difference = partial_difference
        self.mode = mode

    def __call__(self,
                 angles=0, 
                 mus=1, 
                 index_pd_angle=None):
        """
        The output is set on the same device with the angles
        """

        # Number of angles
        if isinstance(angles,list) and len(angles) == 0:
            if isinstance(mus, int) or isinstance(mus, float):
                nMatrices_ = 1
                nDims_ = 1
            elif not torch.is_tensor(mus):
                nMatrices_ = len(mus)
                nDims_ = len(mus[0])
            else:
                nMatrices_ = mus.size(0)
                nDims_ = mus.size(1)
            nAngles_ = int(nDims_*(nDims_-1)/2)
            angles = torch.zeros(nMatrices_,nAngles_)   
        if isinstance(angles, int) or isinstance(angles, float):
            angle_ = angles
            angles = torch.zeros(1,1,dtype=self.dtype)
            angles[0,0] = angle_
        elif not torch.is_tensor(angles):
            angles = torch.tensor(angles,dtype=self.dtype)
        else:
            angles = angles.to(dtype=self.dtype) 
        nAngles = angles.size(1)
        nMatrices = angles.size(0)

        # Setup at the first call
        if not hasattr(self, 'number_of_dimensions'):
            self.number_of_dimensions = int(1+math.sqrt(1+8*nAngles)/2)
        if self.mode == 'sequential' and not hasattr(self, 'nextangle'):
            self.nextangle = 0
        
        # Number of dimensions
        nDims = self.number_of_dimensions

        # Setup of mus, which is send to the same device with angles
        if isinstance(mus, int) or isinstance(mus, float):
            mu_ = mus
            mus = torch.ones(nMatrices,nDims,dtype=self.dtype,device=angles.device)
            mus = mu_ * mus
        elif not torch.is_tensor(mus): #isinstance(mus, list):
            mus = torch.tensor(mus,dtype=self.dtype,device=angles.device)
        else:
            mus = mus.to(dtype=self.dtype,device=angles.device) 
        if mus.size(0) != nMatrices:
            if mus.size(0) == 1:
                mus = torch.tile(mus, (nMatrices,1))
            else:
                raise ValueError("The number of matrices must be the same between angles and mus.")
        elif mus.size(1) != nDims:
            if mus.size(1) == 1:
                mus = torch.tile(mus, (1,nDims))
            else:
                raise ValueError("The number of dimensions must be the same between angles and mus.")
        
        # Switch mode 
        if self.partial_difference and self.mode == 'normal':
            matrix = self.stepNormal_(angles, mus, index_pd_angle)
        elif self.partial_difference and self.mode == 'sequential':
            matrix = self.stepSequential_(angles, mus, index_pd_angle)
        else:
            matrix = self.stepNormal_(angles, mus, index_pd_angle=None)

        return matrix.clone()
    
    def stepNormal_(self, angles, mus, index_pd_angle):
        is_pd = self.partial_difference
        nDims = self.number_of_dimensions
        nMatrices = angles.size(0)
        matrix = torch.tile(torch.eye(nDims,dtype=self.dtype,device=angles.device), (nMatrices,1, 1))
        # 
        for iMtx in range(nMatrices):
            iAng = 0
            matrix_iMtx = matrix[iMtx]
            angles_iMtx = angles[iMtx]
            mus_iMtx = mus[iMtx]
            for iTop in range(nDims - 1):
                vt = matrix_iMtx[iTop, :]
                for iBtm in range(iTop + 1, nDims):
                    angle = angles_iMtx[iAng]
                    if is_pd and iAng == index_pd_angle:
                        angle = angle + torch.pi / 2.0
                    vb = matrix_iMtx[iBtm, :]
                    vt, vb = rot_(vt, vb, angle)
                    if is_pd and iAng == index_pd_angle:
                        matrix_iMtx = torch.zeros_like(matrix_iMtx)
                    matrix_iMtx[iBtm, :] = vb
                    iAng += 1
                matrix_iMtx[iTop, :] = vt
            matrix[iMtx] = mus_iMtx.view(-1,1) * matrix_iMtx

        return matrix

    def stepSequential_(self, angles, mus, index_pd_angle):
        # Check index_pd_angle
        if index_pd_angle != None and index_pd_angle != self.nextangle:
            raise ValueError("Unable to proceed sequential differentiation. Index = %d is expected, but %d was given." % (self.nextangle, index_pd_angle))
        #
        nDims = self.number_of_dimensions
        nMatrices = angles.size(0)
        matrix = torch.tile(torch.eye(nDims,dtype=self.dtype,device=angles.device), (nMatrices,1, 1))
        if index_pd_angle == None: # Initialization 
            self.matrixpst = torch.tile(torch.eye(nDims,dtype=self.dtype,device=angles.device), (nMatrices,1, 1))
            self.matrixpre = torch.tile(torch.eye(nDims,dtype=self.dtype,device=angles.device), (nMatrices,1, 1))
            for iMtx in range(nMatrices):
                iAng = 0
                matrixpst_iMtx = self.matrixpst[iMtx]
                angles_iMtx = angles[iMtx]
                mus_iMtx = mus[iMtx]
                for iTop in range(nDims - 1):
                    vt = matrixpst_iMtx[iTop, :]
                    for iBtm in range(iTop + 1, nDims):
                        angle = angles_iMtx[iAng]
                        vb = matrixpst_iMtx[iBtm, :]
                        vt, vb = rot_(vt, vb, angle)
                        matrixpst_iMtx[iBtm, :] = vb
                        iAng += 1
                    matrixpst_iMtx[iTop, :] = vt
                matrix[iMtx] = mus_iMtx.view(-1,1) * matrixpst_iMtx
                self.matrixpst[iMtx] = matrixpst_iMtx
            self.nextangle = 0
        else: # Sequential differentiation
            for iMtx in range(nMatrices):
                matrixrev = torch.eye(nDims,dtype=self.dtype,device=angles.device)
                matrixdif = torch.zeros(nDims,nDims,dtype=self.dtype,device=angles.device)
                iAng = 0
                matrixpst_iMtx = self.matrixpst[iMtx]
                matrixpre_iMtx = self.matrixpre[iMtx]
                angles_iMtx = angles[iMtx]
                mus_iMtx = mus[iMtx]
                for iTop in range(nDims - 1):
                    rt = matrixrev[iTop, :]
                    dt = torch.zeros(nDims,dtype=self.dtype,device=angles.device)
                    dt[iTop] = 1
                    for iBtm in range(iTop + 1, nDims):
                        if iAng == index_pd_angle:
                            angle = angles_iMtx[iAng]
                            rb = matrixrev[iBtm, :]
                            rt, rb = rot_(rt, rb, -angle)
                            matrixrev[iTop, :] = rt
                            matrixrev[iBtm, :] = rb
                            db = torch.zeros(nDims,dtype=self.dtype,device=angles.device)
                            db[iBtm] = 1
                            dangle = angle + torch.pi / 2.0
                            dt, db = rot_(dt, db, dangle)
                            matrixdif[iTop, :] = dt
                            matrixdif[iBtm, :] = db
                            matrixpst_iMtx = torch.matmul(matrixpst_iMtx, matrixrev)
                            matrix[iMtx] = torch.matmul(torch.matmul(matrixpst_iMtx, matrixdif), matrixpre_iMtx)
                            matrixpre_iMtx = torch.matmul(torch.transpose(matrixrev,0,1), matrixpre_iMtx)
                        iAng += 1
                matrix[iMtx] = mus_iMtx.view(-1,1) * matrix[iMtx]
                self.matrixpre[iMtx] = matrixpre_iMtx
                self.matrixpst[iMtx] = matrixpst_iMtx
            self.nextangle += 1

        return matrix

File: code/+tansacnet/+utility/OrthonormalMatrixGenerationSystem.m
classdef OrthonormalMatrixGenerationSystem < matlab.System %#codegen
    %ORTHONORMALMATRIXGENERATIONSYSTEM Orthonormal matrix generator
    %
    % Requirements: MATLAB R2015b
    %
    % Copyright (c) 2014-2024, Shogo MURAMATSU, Yasas GODAGE
    %
    % All rights reserved.
    %
    % Contact address: Shogo MURAMATSU, Yasas GODAGE
    %                Faculty of Engineering, Niigata University,
    %                8050 2-no-cho Ikarashi, Nishi-ku,
    %                Niigata, 950-2181, JAPAN
    %
    % http://msiplab.eng.niigata-u.ac.jp/
    %
    
    properties (Nontunable)
        PartialDifference = 'off'
    end
    
    properties (Hidden, Transient)
        PartialDifferenceSet = ...
            matlab.system.StringSet({'on','off','sequential'});
        DeviceSet = ...
            matlab.system.StringSet({'cpu','cuda'});
        DTypeSet = ...
            matlab.system.StringSet({'single','double'});
    end
    
    properties
        NumberOfDimensions
        Device = 'cuda'
        DType = 'double'
    end
    
    properties (Access = private)
        matrixpst
        matrixpre
        nextangle
    end
    
    methods
        function obj = OrthonormalMatrixGenerationSystem(varargin)
            % Support name-value pair arguments
            setProperties(obj,nargin,varargin{:});
        end
    end
    
    methods (Access = protected)
        
        function s = saveObjectImpl(obj)
            s = saveObjectImpl@matlab.System(obj);
            s.NumberOfDimensions = obj.NumberOfDimensions;
            s.PartialDifference = obj.PartialDifference;
            s.Device = obj.Device;
            s.DType = obj.DType;
        end
        
        function loadObjectImpl(obj,s,wasLocked)
            if isfield(s,'PartialDifference')
                obj.PartialDifference = s.PartialDifference;
            else
                obj.PartialDifference = 'off';
            end
            obj.DType = s.DType;
            obj.Device = s.Device;            
            obj.NumberOfDimensions = s.NumberOfDimensions;
            loadObjectImpl@matlab.System(obj,s,wasLocked);
        end
        
        function setupImpl(obj,angles,~,~)
            if isempty(obj.NumberOfDimensions)
                obj.NumberOfDimensions = (1+sqrt(1+8*size(angles,1)))/2;
            end
            if strcmp(obj.PartialDifference,'sequential')
                obj.nextangle = uint32(0);            
            end
        end
        
        function resetImpl(obj)
            obj.nextangle = uint32(0);
        end

        function validateInputsImpl(obj,angles,mus,~)
            % Check device of angles
            if obj.Device == "cuda" && ~isgpuarray(angles)
                errID = 'LSUN:DeviceMismatch';
                msg = 'ANGLES should be gpuArray';
                throw(MException(errID,msg));
            elseif obj.Device == "cpu" && isgpuarray(angles)
                errID = 'LSUN:DeviceMismatch';
                msg = 'ANGLES should be on CPU';
                throw(MException(errID,msg));
            end

            % Check device of mus
            if obj.Device == "cuda" && ~isgpuarray(mus)
                errID = 'LSUN:DeviceMismatch';
                msg = 'MUS should be gpuArray';
                throw(MException(errID,msg));
            elseif obj.Device == "cpu" && isgpuarray(mus)
                errID = 'LSUN:DeviceMismatch';
                msg = 'MUS should be on CPU';
                throw(MException(errID,msg));
            end

            % Check dtype of angles
            if isgpuarray(angles)
                angles = gather(angles);
            end
            if ~strcmp(obj.DType,class(angles))
                errID = 'LSUN:DTypeMismatch';
                msg = char("ANGLES should be " + obj.DType);
                throw(MException(errID,msg));
            end

            % Check dtype of angles
            if isgpuarray(mus)
                mus = gather(mus);
            end
            if ~strcmp(obj.DType,class(mus))
                errID = 'LSUN:DTypeMismatch';
                msg = char("MUS should be " + obj.DType);
                throw(MException(errID,msg));
            end

            if ~isempty(mus) && any(abs(mus(:))~=1)
                error('All entries of mus must be 1 or -1.');
            end
        end
        
        function matrix = stepImpl(obj,angles,mus,pdAng)
            device_ = obj.Device;
            dtype_ = obj.DType;

            if nargin < 4
                pdAng = 0;
            end
            
            if isempty(angles)
                if isvector(mus) % Single case
                    matrix = diag(mus); 
                else % Multiple case
                    if device_ == "cuda"
                        matrix = zeros(size(mus,1),size(mus,1),size(mus,2),dtype_,"gpuArray");
                    else
                        matrix = zeros(size(mus,1),size(mus,1),size(mus,2),dtype_);
                    end
                    for idx = 1:size(mus,2)
                        matrix(:,:,idx) = diag(mus(:,idx));
                    end
                end
            elseif strcmp(obj.PartialDifference,'sequential')
                % Sequential mode
                matrix = obj.stepSequential_(angles,mus,pdAng);
            else
                % Normal mode
                matrix = obj.stepNormal_(angles,mus,pdAng);
            end
        end
        
        function N = getNumInputsImpl(obj)
            if strcmp(obj.PartialDifference,'on') || ...
                    strcmp(obj.PartialDifference,'sequential')
                N = 3;
            else
                N = 2;
            end
        end
        
        function N = getNumOutputsImpl(~)
            N = 1;
        end
    end
    
    methods (Access = private)
        
        function matrix = stepNormal_(obj,angles,mus,pdAng)
            device_ = obj.Device;
            dtype_ = obj.DType;
            nDim_ = obj.NumberOfDimensions;
            nMatrices_ = size(angles,2);
            if device_ == "cuda"
                eye_ = eye(nDim_,dtype_,"gpuArray");
            else
                eye_ = eye(nDim_,dtype_);
            end
            matrix = repmat(eye_,[1 1 nMatrices_]); 
            if isrow(mus)
                mus = mus.';
            end
            for iMtx = 1:nMatrices_
                iAng = 1;
                for iTop=1:nDim_-1
                    vt = matrix(iTop,:,iMtx);
                    for iBtm=iTop+1:nDim_
                        angle = angles(iAng,iMtx);
                        if iAng == pdAng
                            angle = angle + pi/2;
                        end
                        vb = matrix(iBtm,:,iMtx);
                        [vt,vb] = obj.rot_(vt,vb,angle);
                        if iAng == pdAng
                            matrix(:,:,iMtx) = cast(0,dtype_)*matrix(:,:,iMtx);
                        end
                        matrix(iBtm,:,iMtx) = vb;
                        %
                        iAng = iAng + 1;
                    end
                    matrix(iTop,:,iMtx) = vt;
                end
                if iscolumn(mus) || isscalar(mus)
                    matrix(:,:,iMtx) = mus.*matrix(:,:,iMtx);
                else
                    matrix(:,:,iMtx) = mus(:,iMtx).*matrix(:,:,iMtx);
                end
            end
        end

        function matrix = stepSequential_(obj,angles,mus,pdAng)
            device_ = obj.Device;
            dtype_ = obj.DType;
            % Check pdAng
            if pdAng ~= obj.nextangle
                error("Unable to proceed sequential differentiation. Index = %d is expected, but %d was given.", obj.nextangle, pdAng);
            end
            %
            nDim_ = obj.NumberOfDimensions;
            nMatrices_ = size(angles,2);
            if device_ == "cuda"
                eye_ = eye(nDim_,dtype_,"gpuArray"); 
                zeros_ = zeros(nDim_,dtype_,"gpuArray"); 
            else
                eye_ = eye(nDim_,dtype_); 
                zeros_ = zeros(nDim_,dtype_); 
            end
            matrix = repmat(eye_,[1 1 nMatrices_]);
            if isrow(mus)
                mus = mus.';
            end
            if pdAng < 1 % Initialization
                obj.matrixpst = repmat(eye_,[1 1 nMatrices_]); 
                obj.matrixpre = repmat(eye_,[1 1 nMatrices_]); 
                %
                for iMtx = 1:nMatrices_
                    iAng = 1;
                    for iTop=1:nDim_-1
                        vt = obj.matrixpst(iTop,:,iMtx);
                        for iBtm=iTop+1:nDim_
                            angle = angles(iAng,iMtx);
                            vb = obj.matrixpst(iBtm,:,iMtx);
                            [vt,vb] = obj.rot_(vt,vb,angle);
                            obj.matrixpst(iBtm,:,iMtx) = vb;
                            iAng = iAng + 1;
                        end
                        obj.matrixpst(iTop,:,iMtx) = vt;
                    end
                    if iscolumn(mus)
                        matrix(:,:,iMtx) = mus.*obj.matrixpst(:,:,iMtx);
                    else
                        matrix(:,:,iMtx) = mus(:,iMtx).*obj.matrixpst(:,:,iMtx);
                    end
                end
                obj.nextangle = uint32(1);
            else % Sequential differentiation
                if device_ == "cuda"
                    zeros_1xD_ = zeros(1,nDim_,dtype_,"gpuArray");
                else
                    zeros_1xD_ = zeros(1,nDim_,dtype_);
                end
                %
                %matrix = 1;
                for iMtx = 1:nMatrices_
                    matrixrev = eye_; 
                    matrixdif = zeros_; 
                    %
                    iAng = 1;
                    for iTop=1:nDim_-1
                        rt = matrixrev(iTop,:);
                        dt = zeros_1xD_; 
                        dt(iTop) = 1;
                        for iBtm=iTop+1:nDim_
                            if iAng == pdAng
                                angle = angles(iAng,iMtx);
                                %
                                rb = matrixrev(iBtm,:);
                                [rt,rb] = obj.rot_(rt,rb,-angle);
                                matrixrev(iTop,:) = rt;
                                matrixrev(iBtm,:) = rb;
                                %
                                db = zeros_1xD_; 
                                db(iBtm) = 1;
                                dangle = angle + pi/2;
                                [dt,db] = obj.rot_(dt,db,dangle);
                                matrixdif(iTop,:) = dt;
                                matrixdif(iBtm,:) = db;
                                %
                                obj.matrixpst(:,:,iMtx) = obj.matrixpst(:,:,iMtx)*matrixrev;
                                matrix(:,:,iMtx) = obj.matrixpst(:,:,iMtx)*matrixdif*obj.matrixpre(:,:,iMtx);
                                obj.matrixpre(:,:,iMtx) = matrixrev.'*obj.matrixpre(:,:,iMtx);
                            end
                            iAng = iAng + 1;
                        end
                    end
                    if iscolumn(mus)
                        matrix(:,:,iMtx) = mus.*matrix(:,:,iMtx);
                    else
                        matrix(:,:,iMtx) = mus(:,iMtx).*matrix(:,:,iMtx);
                    end                    
                end
                obj.nextangle = obj.nextangle + 1;
            end
        end
    end
    
    methods (Static, Access = private)
        function [vt,vb] = rot_(vt,vb,angle)
            c = cos(angle);
            s = sin(angle);
            %
            u  = s*(vt + vb);
            vt = (c + s)*vt;
            vb = (c - s)*vb;
            vt = vt - u;
            vb = vb + u;
        end
    end
end
